// nativescript/lib/common/mobile/ios/simulator/ios-simulator-log-provider.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
class IOSSimulatorLogProvider {
    constructor($iOSSimResolver, $deviceLogProvider, $devicePlatformsConstants, $logger, $processService) {
        this.$iOSSimResolver = $iOSSimResolver;
        this.$deviceLogProvider = $deviceLogProvider;
        this.$devicePlatformsConstants = $devicePlatformsConstants;
        this.$logger = $logger;
        this.$processService = $processService;
        this.simulatorsLoggingEnabled = {};
    }
    startLogProcess(deviceIdentifier) {
        if (!this.simulatorsLoggingEnabled[deviceIdentifier]) {
            const deviceLogChildProcess = this.$iOSSimResolver.iOSSim.getDeviceLogProcess(deviceIdentifier, 'senderImagePath contains "NativeScript"');
            const action = (data) => {
                this.$deviceLogProvider.logData(data.toString(), this.$devicePlatformsConstants.iOS, deviceIdentifier);
            };
            if (deviceLogChildProcess) {
                deviceLogChildProcess.once("close", () => {
                    this.simulatorsLoggingEnabled[deviceIdentifier] = false;
                });
                deviceLogChildProcess.once("error", (err) => {
                    this.$logger.trace(`Error is thrown for device with identifier ${deviceIdentifier}. More info: ${err.message}.`);
                    this.simulatorsLoggingEnabled[deviceIdentifier] = false;
                });
            }
            if (deviceLogChildProcess.stdout) {
                deviceLogChildProcess.stdout.on("data", action);
            }
            if (deviceLogChildProcess.stderr) {
                deviceLogChildProcess.stderr.on("data", action);
            }
            this.$processService.attachToProcessExitSignals(this, deviceLogChildProcess.kill);
            this.simulatorsLoggingEnabled[deviceIdentifier] = true;
        }
    }
}
exports.IOSSimulatorLogProvider = IOSSimulatorLogProvider;
$injector.register("iOSSimulatorLogProvider", IOSSimulatorLogProvider);
